### PR TITLE
revert: StartSpanFromContext for Tracer interface

### PR DESCRIPTION
**Pull request by felixge**
Remove StartSpanFromContext method from ddtrace.Tracer interface to
avoid breaking users that implement this interface.

Revert tracer.StartSpanFromContext to its old behavior of inherting the
span contained in the passed context. This makes sure that users who
write tests that rely on the default NoopTracer see the same behavior as
previously.

This reverts commits 5f95c9f and 6661a9d.

### CODE REVIEW

File: ddtrace/tracer/context.go
Comment (Julio-Guerra) [inline ddtrace/tracer/context.go:42]: Note that a separate `interface { StartSpanFromContext(...) }` could do the job too if you want to keep this method in `tracer` and call it only in that case. A `startSpanFromContext()` function could hold the logic below, also usable from your method:

```go
type spanStarter interface {
  StartSpanFromContext(ctx context.Context, operationName string, opts ...StartSpanOption) (Span, context.Context)
}

tracer := internal.GetGlobalTracer()
if starter, ok := tracer.( .StartSpanFromContext(ctx, operationName, opts...); ok {
  return starter.startSpanFromContext(ctx, operationName, opts...)
} else {
  return startSpanFromContext(ctx, operationName, opts...)
}
```

But now that I wrote it, maybe it's worst compared to the simple `Context` field you added to the start options :-)

File: ddtrace/tracer/tracer.go
Comment (Julio-Guerra) [inline ddtrace/tracer/tracer.go:330]:
```suggestion
	// The default pprof context is taken from the start options and is
	// not nil when using StartSpanFromContext()
	pprofContext := opts.Context
```

**Review by Julio-Guerra — CHANGES_REQUESTED** [review]
I added suggestions and a few comments.

The reasons why I don't approve it is that if such backward compatibility of the `Tracer` interface type is required, then we should introduce a simple test of it by simply having in a test file a type implementing the current methods of `Tracer`, along with a simple `var _ ddtracer.Tracer = myCustomImplementation{}` so that it breaks if we change it - or a run-time check such as `require.Implements(t, (*ddtrace.Tracer)(nil), &myCustomImplementation{})`

WDYT?

File: ddtrace/ddtrace.go
Comment (felixge) [reply ddtrace/ddtrace.go:130]: > you could also make this new field private

Go's visibility rules won't allow me to make this field private. This struct is declared in pkg `ddtrace` but I need to access it in `ddtrace/tracer`.

File: ddtrace/tracer/context.go
Comment (felixge) [reply ddtrace/tracer/context.go:42]: Interesting idea. But yeah in practice I think it's going to add more complexity/indirection than this patch. So I think I'd prefer to keep going in the current direction.

**Review by felixge — COMMENTED** [review]
> The reasons why I don't approve it is that if such backward compatibility of the Tracer interface type is required, then we should introduce a simple test of it by simply having in a test file a type implementing the current methods of Tracer, along with a simple var _ ddtracer.Tracer = myCustomImplementation{} so that it breaks if we change it - or a run-time check such as require.Implements(t, (*ddtrace.Tracer)(nil), &myCustomImplementation{})

We already have such a test in the form of the `mocktracer` and `NoopTracer`. Both of them broke when adding the new interface method. The problem was just that I didn't realize it was a public interface.

File: ddtrace/ddtrace.go
Comment (felixge) [reply ddtrace/ddtrace.go]: e90da29bf154428ac14e4bf1a010dffb585b0595

File: ddtrace/tracer/tracer.go
Comment (felixge) [reply ddtrace/tracer/tracer.go:332]: e90da29bf154428ac14e4bf1a010dffb585b0595

**Review by Julio-Guerra — APPROVED** [review]
All good then, thanks for your replies :-)

**Pull request merged**
